Split imports.js into seperate modules

aidanharris opened the pull request:
There's a fair bit of duplicated code at the top of each module that could be significantly reduced by making a common 'provider' module that each provider includes.

aidanharris committed:
Add providers to config.json

aidanharris committed:
Modularise kat and torrentproject providers. There's a lot of code duplication between each module which could be minimised by creating a common provider module that every provider includes

OmgImAlexis commented on background/imports.js:
```diff
@@ -1,141 +1,14 @@
-        });
+for(var provider in nconf.get('providers')) {
+    if(nconf.get('providers:' + provider + ':enabled')) {
+        console.log('Loading provider ' + provider); // To Do: Add proper logging here
```
Use [bunyan](https://www.npmjs.com/package/bunyan) here for logging.

OmgImAlexis commented on background/providers/kat.js:
```diff
@@ -0,0 +1,103 @@
+var nconf = require('nconf');
+var mongoose = require('mongoose');
+var Category = require('../../app/models/Category.js');
+var Torrent = require('../../app/models/Torrent.js');
```
We should look into using the same thing I did for the app.js file. This allows us to have /app as a relative root.
````
require('app-module-path').addPath(__dirname + '/app');
````

OmgImAlexis commented on background/providers/kat.js:
```diff
@@ -0,0 +1,103 @@
+
+nconf.use('memory');
+nconf.argv().env('__').file({
+    file: path.resolve(__dirname + '/../../config.json')
```
Same goes for here and really we should look at adding nconf to the app global so we don't have to re-init it.

OmgImAlexis commented on background/providers/kat.js:
```diff
@@ -0,0 +1,103 @@
+        break;
+    default:
+        duration = Number(nconf.get('providers:kat:config:duration'));
+}
```
This is really smart.

OmgImAlexis commented on background/providers/kat.js:
```diff
@@ -0,0 +1,103 @@
+                    lines.forEach(function (line) {
+                        line = line.split('|');
+                        console.log('Processing ' + line[1]);
+                        Category.findOne({
```
I think we should try and add a function to the "generic" provider that will then be able to be extended using node's new Classes in ES6.

OmgImAlexis commented on background/providers/kat.js:
```diff
@@ -0,0 +1,103 @@
+
+                });
+            } else {
+                console.log('Kat API key is invalid');
```
We should change these all over to bunyan.

OmgImAlexis commented:
If you want to make those changes I'll merge in, glad to see someone else helping out with this. :+1:

aidanharris commented on background/providers/torrentproject.js:
```diff
@@ -0,0 +1,99 @@
+});
+
+if(nconf.get('database:mongodb:enabled')){
+    mongoose.connect('mongodb://' + nconf.get('database:mongodb:host') + ':' + nconf.get('database:mongodb:port') + '/' + nconf.get('database:mongodb:collection'), function(err){
```
What's this supposed to do? On my computer (Windows 10, Node v5.5.0 mongoose connects fine but `err` still exists with the following error:

```JavaScript
{ [Error: Trying to open unclosed connection.] state: 2 }
```

Can we ignore this error since the database connection is already open or should we be using `createConnection` instead of `connect`?

OmgImAlexis replied:
You're loading the files into a file with a connection already established. We can just remove that from the providers and keep it in the import file so we have a single connection shared between all providers.

aidanharris committed:
Ignore already open connections to MongoDB and don't re-initialise nconf in kat and torrentproject modules

aidanharris committed:
Configure Bunyan to enable logging in imports.js and providers

aidanharris committed:
Configure Bunyan to enable logging in imports.js and providers

aidanharris committed:
Configure Bunyan to enable logging in imports.js and providers

aidanharris committed:
Replace console.foo with Bunyan and format it appropriately

aidanharris committed:
Establish a mongoose connection once in import.js instead of in each provider module

aidanharris committed:
Move logging to a seperate file ('logging.js')

aidanharris committed:
Make providers object-oriented

        * Providers now inherit from a 'Provider' class
        * Logging has been moved to a dedicated file ('logging.js') that is required from files
        * Whitelist / blacklist functionality has been added
        * Categories are now added to the database if they aren't there already.
        * Some bugs may exist in my implementation, most notably #24

aidanharris commented:
Is this project still active? If so, I'd like to continue working on it. This pull request is full of a lot of changes so I think it would be best if I closed this pull request, reviewed the code and split it up into multiple pull requests. Any thoughts?

stephen304 commented:
I haven't really been active at all, though I'd like to let people continue working on it if they want. It looks like OmgImAlexis has been doing quite a lot, I'm not sure if adding more contributors would be helpful, thoughts @OmgImAlexis ?

OmgImAlexis commented:
I'm fine with it, I've had a lot of personal stuff going on as I've just moved from Australia to the UK. Once I get my server and stuff all sorted I'll be back to working on this.

I'll have a look through and see if this is worth merging or if anything needs to be changed.

OmgImAlexis commented on background/logging.js:
```diff
@@ -0,0 +1,32 @@
+
+formatBunyanInfoLog = function() {};
+formatBunyanInfoLog.prototype.write = function (rec) {
+    console.log('[%s] %s: %s',
```
Would it be possible to have this as an option so people can change the logging style if they wanted to? We should also probably document this in a wiki or something so if anyone wants to use an external logger such as graylog or ELK they can easily import it.

OmgImAlexis commented on background/logging.js:
```diff
@@ -0,0 +1,32 @@
+};
+
+logger = bunyan.createLogger({
+    name: 'Bitcannon',
```
This should also be showing that the logs are coming from the web interface since we wanted to modalise this and break it into separate processes.

aidanharris replied:
How should logging be handled?

We could allow the log level and format strings to be customised in the config.json and perhaps allow external loggers via a logging class / module of some sort that can be swapped out for others as long as they conform to a common interface.

>This should also be showing that the logs are coming from the web interface since we wanted to modalise this and break it into separate processes.

Do you mean show logs from the web interface (server) and imports at the same time? How could we go about doing this?

OmgImAlexis replied:
Would it be possible for the log.warn, etc. to add a custom field? We could then have `threadType` and have that set as webui, importer, etc.

aidanharris replied:
What about something like this: https://github.com/aidanharris/bitcannon-web/commit/7df89c52a5588a25be30e4368c842e42cd78aebc

I've created a logging class and it's initialised with the thing it's logging e.g `new logger('webui');` and then this is printed in the output of the logging.

OmgImAlexis commented on background/providers/kat.js:
```diff
@@ -0,0 +1,69 @@
+
+class Kat extends Provider {
+    run() {
+        https.get('https://kat.cr/api/get_dump/hourly/?userhash=' + nconf.get('providers:kat:config:apiKey'), function (res) {
```
This should also be using a value from the config.json since people may need to use a proxy address if kat.cr is blocked.

OmgImAlexis commented on background/providers/provider.js:
```diff
@@ -0,0 +1,186 @@
+    }
+
+    // Add a torrent to the database
+    static addTorrent(title, aliases, size, details, swarm, lastmod, imported, infoHash) {
```
We should also add a issue about adding torrent files themselves either saving into a directory or as a base64 string in the database since this would allow us to work with private torrents that don't use magnets.

OmgImAlexis commented on background/providers/provider.js:
```diff
@@ -0,0 +1,186 @@
+    // Add a torrent to the database
+    static addTorrent(title, aliases, size, details, swarm, lastmod, imported, infoHash) {
+        // Validate Data
+        if(typeof(title) !== 'string' || typeof(infoHash) !== 'string') {
```
The same goes for here since if it's a private tracker it may not have a infoHash/magnet link from the API.

OmgImAlexis commented on config.json:
```diff
@@ -14,22 +14,38 @@
+        }
+    },
     "session": {
         "secret": "asdasdasdasd43jn5kj2n43k"
```
This should be "" as default and be created on first run otherwise all installs will be using the same session secret. I can't remember if we do that or not.

OmgImAlexis commented:
If we address those issues then it should be fine to merge.

aidanharris committed:
Remove default session secret from config.json

aidanharris committed:
Fix broken relative require path in latest version of Node.js (v6.2.1)

aidanharris committed:
Remove hardcoded Kat URL in favour of configuration via config.json

aidanharris committed:
Remove hardcoded Torrentproject URL in favour of configuration via config.json

aidanharris committed:
Add note about storing torrent files (either in a directory or in the database) if we don't have a magnet link or infohash

OmgImAlexis commented on config.json:
```diff
@@ -34,7 +34,8 @@
                 "startup": true,
                 "duration": "@hourly",
-                "apiKey": "fknelkn32l4kn23l4kn23lk4n"
+                "apiKey": "fknelkn32l4kn23l4kn23lk4n",
```
We should also have this set as "".

aidanharris committed:
Exception handling when loading providers

aidanharris committed:
Logging Classes

aidanharris committed:
Add threadType to Bunyan

aidanharris committed:
Add Vagrantfile

aidanharris committed:
Hopefully improve import performance

  * Updated ECMAScript version to 6 in .jshintrc
  * Added required and unique attributes to Mongoose models and defaults for lastmod, imported and torrentCount
  * Added the Node.js cluster module to take advantage of multi-core systems. No doubt there is potentially further optimisation that can be achieved here.
  * Removed Mongoose from the imports and instead directly make use of the native MongoDB driver, it's so much faster than Mongoose, especially when adding multiple items to the database at one time.
  * Added JSDoc (http://usejsdoc.org) comments to functions and classes. Documentation can be generated as follows:
    - `npm install -g jsdoc`
    - `jsdoc index.js background/*.js background/providers/*.js background/logging/*.js`
    - To Do: Add a grunt plugin to generate documentation.

aidanharris committed:
Remove Kat API Key

aidanharris committed:
Update Vagrantfile

aidanharris committed:
Update Torrentproject URL

I forgot to change the URL before pushing :(

aidanharris committed:
Added a helper module to return the http or https module depending on a URL and disabled the Kat provider by default since it needs an api key

aidanharris committed:
Thread scheduling for providers

aidanharris committed:
Added xml2js to package.json

aidanharris committed:
Added parse-torrent to package.json

aidanharris committed:
WIP, Quick & Dirty RSS Parser

aidanharris committed:
RSS Feed items with the mimetype application/x-bittorrent are torrents

aidanharris committed:
Remove redundant console.log

aidanharris committed:
Remove redundant config options and generate session secret on-demand

aidanharris committed:
Remove unusable providers

OmgImAlexis commented:
I haven't had a chance to look at this since I stopped working on it a while ago. I'll try and review today and see if we can start work on the project again. Thank you @aidanharris for all the work.

OmgImAlexis commented on background/imports.js:
```diff
@@ -1,141 +1,166 @@
+ * process multiple archives at a time. However if this code is io bound rather than CPU bound it
+ * might make no difference at all. Further testing is required.
+ */
+if(cluster.isMaster) {
```
This may all need to be removed as we should be moving to using ES6 and imports instead of require.

aidanharris replied:
How are we going to switch to using imports? Will we need to add [babel](https://babeljs.io/docs/usage/cli/#babel-node)?

OmgImAlexis replied:
Yes.

OmgImAlexis commented on background/imports.js:
```diff
@@ -1,141 +1,166 @@
+        nconf.get('torrents:blacklist:enabled')
+    ) {
+        log.error('You cannot use the whitelist and the blacklist at the same time!');
+        process.exit(1);
```
I'd like to switch from jshint to [xo](https://github.com/sindresorhus/xo) which requires this file to have `#!/usr/bin/env node` at the start if we use `process.exit()`. The better idea would be to exit using `new Error()`.

OmgImAlexis commented on background/imports.js:
```diff
@@ -1,141 +1,166 @@
         });
+    } else {
+        log.warn('No database is enabled, please check your config.json');
+        process.exit(1);
```
Same goes for here about the `process.exit()`.

OmgImAlexis commented on background/providers/provider.js:
```diff
@@ -0,0 +1,198 @@
+/**
+ * @file Provides a baseclass for archive providers
```
Maybe this would link to a wiki page explaining what an "archive providers" is.

OmgImAlexis commented on background/providers/rss.js:
```diff
@@ -0,0 +1,363 @@
+      parseTorrent.remote(url,
+        function (err, parsedTorrent) {
+          if (err) {
+            log.info(err);
```
This should all be using template strings when we move to ES6.

OmgImAlexis commented on config.json:
```diff
@@ -1,42 +1,54 @@
+    }
+  },
+  "session": {
+    "secret": ""
```
I'd like to move to JWT if possible so we can move the UI to using Vue as that's what I'm using on all other projects. It'd allow the server as well as the background processes to run on a seedbox, etc. and have the ui bundled as an Electron app or even just hosted locally via Apache, Nginx, Caddy, etc.

aidanharris replied:
Do you think it would be a good idea to split Bitcannon into separate repositories (webui, api and importer daemon)?

OmgImAlexis replied:
Sounds like a good idea. I was also thinking maybe we should rebrand as a different app since this isn't really like the original and the original guy that made this isn't really involved. Thoughts?

aidanharris replied:
I have no problem with that. If we're going to re-brand we'll need a new name (could just use something generic for now). We should also try to come up with a roadmap so we can see all of the things that need to be done and track issues. If @stephen304 doesn't have any objections it perhaps wouldn't be a bad idea to update the README to tell people development has halted and continued elsewhere.

OmgImAlexis replied:
The dev won't be halted here and moved we'd just rename this repo since the "real" Bitcannon is still under stephen's account and written in Golang.

aidanharris replied:
If you aren't busy, I'm in the [gitter](https://gitter.im/bitcannon-org) chat if you'd like to talk about things. I have some ideas for how we can improve things.

aidanharris committed:
Style updates to imports and minor tweaks

OmgImAlexis commented on Dockerfile.imports:
```diff
@@ -0,0 +1,11 @@
+
+WORKDIR /var/lib/bitcannon/imports
+
+RUN yarn install && \
```
Use `yarn install --production`.

OmgImAlexis replied:
You should also be using `ENV NODE_ENV "production"`

OmgImAlexis commented on background/imports.js:
```diff
@@ -6,36 +6,38 @@
  nconf.use('memory');
  nconf.argv().env('__').file({
-     file: path.resolve(__dirname + '/../config.json')
+     file: path.resolve(`${__dirname}/../config.json`)
```
`path.join()` should be used here incase of windows issues with slashes.
